coreneuron/mechanism/mechanism.hpp

```diff
@@ -76,6 +79,50 @@ struct NetSendBuffer_t {
     int _cnt;
     int _size;       /* capacity */
     int reallocated; /* if buffer resized/reallocated, needs to be copy to cpu */
+
+    NetSendBuffer_t(int size) : _size(size) {
+        _cnt = 0;
+
+        _sendtype = (int*)ecalloc_align(_size, sizeof(int));
+        _vdata_index = (int*)ecalloc_align(_size, sizeof(int));
+        _pnt_index = (int*)ecalloc_align(_size, sizeof(int));
+        _weight_index = (int*)ecalloc_align(_size, sizeof(int));
+        // when == 1, NetReceiveBuffer_t is newly allocated (i.e. we need to free previous copy
+        // and recopy new data
+        reallocated = 1;
+        _nsb_t = (double*)ecalloc_align(_size, sizeof(double));
+        _nsb_flag = (double*)ecalloc_align(_size, sizeof(double));
+    }
+
+    ~NetSendBuffer_t() {
+        free(_sendtype);
+        free(_vdata_index);
+        free(_pnt_index);
+        free(_weight_index);
+        free(_nsb_t);
+        free(_nsb_flag);
+    }
+
+    void grow() {
+        int new_size = _size * 2;
+        grow_buf(&_sendtype, _size, new_size);
+        grow_buf(&_vdata_index, _size, new_size);
+        grow_buf(&_pnt_index, _size, new_size);
+        grow_buf(&_weight_index, _size, new_size);
+        grow_buf(&_nsb_t, _size, new_size);
+        grow_buf(&_nsb_flag, _size, new_size);
+    }
+
+    private:
+        template <typename T>
+        void grow_buf(T** buf, int size, int new_size) {
```

Comment:
Would `realloc` be sufficient here? (I haven't checked but do we have `realloc_align` or something already?)

Reply:
No, we don't have it. I could of course implement that. there is no `_aligned_realloc` in stdlib, I checked that. Hence I had to do the allocation, and `memcpy` myself. But on the other hand, we still have the plan to replace `memory.h` by a newer implementation that uses proper C++ allocators. Not sure how concrete this plan is. If yes, then this is good enough, if not I can add the generic implementation of aligned realloc.

Reply:
That's true. And I think that all aligned* should go away now (not really required nowadays).